### PR TITLE
Analysis Requests are never transitioned to assigned/unassigned

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

This Pull Requests removes `AnalysisRequest` portal type from `bika_worksheetanalysis_workflow` workflow. This workflow only makes sense for analyses (which are the only objects that can be assigned/unassigned to a Worksheet and to an analyst therefore). It was only used in AnalysisRequest for filtering in lists, but was not working because ARs were never transitioned to `assigned` or `unassigned`. With this PR, this has been addressed with a new index and metadata in AR's catalog.

Thus, this Pull Request contains the following:
- Removed `AnalysisRequest` portal type from `bika_worksheetanalysis_workflow` 
- New index and metadata column `assigned_state` in `analysisrequest_catalog_listing`
- Each time an Analysis is assigned/unassigned to a worksheet, the associated Analysis Request is reindexed (thus, `assigned_state` index+metadata updated)

**Additional note**: This Pull Request does not consider if an AnalysisRequest in, e.g. `verified` state, should be displayed when the filter `assigned` is used or not. This is considered in https://github.com/senaite/senaite.core/pull/627

## Current behavior before PR

Analysis Requests cannot be filtered by assigned/unassigned states.

## Desired behavior after PR is merged

Analysis Requests can be filtered by assigned/unassigned states.

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/catalog/indexers/analysisrequest.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    analyses the analysisrequest contains. Return `unassigned` if the Analysis
+    Request does not contain any analysis or if has at least one in `unassigned`
+    state. Otherwise, returns `assigned`"""
+    analyses = instance.getAnalyses()
+    if not analyses:
+        return 'unassigned'
+
+    for analysis in analyses:
+        analysis = api.get_object(analysis)
+        state = getCurrentState(analysis, 'worksheetanalysis_review_state')
```
Comment (ramonski) [inline bika/lims/catalog/indexers/analysisrequest.py]: please use `api.get_workflow_status_of` instead, then you don't have to ensure the object in line 25 first: https://github.com/senaite/senaite.core/blob/master/bika/lims/api.py#L740

File: RELEASE_NOTES.rst [hunk 2]
```diff
@@ -1,6 +1,16 @@
 Release notes
 =============
 
+Update from 1.2.1 to 1.2.2
+--------------------------
```
Comment (ramonski) [inline RELEASE_NOTES.rst]: Needs to be now 1.2.2 to 1.2.3

File: bika/lims/upgrade/v01_02_002.py [hunk 3]
```diff
@@ -47,6 +50,10 @@ def upgrade(tool):
     # section from Dashboard
     add_sample_section_in_dashboard(portal)
 
+    # Unbound the worksheetanalysis_workflow from Analysis Requests and add a
+    # FieldIndex 'assigned_state' in AR's catalog (for its use on searches)
+    fix_assign_analysis_requests(portal, ut)
```
Comment (ramonski) [inline bika/lims/upgrade/v01_02_002.py]: This upgrade step needs to go to 1.2.3

File: bika/lims/workflow/analysis/events.py [hunk 4]
```diff
@@ -187,8 +201,11 @@ def after_attach(obj):
     _reindex_request(obj)
 
 
-def _reindex_request(obj):
-    if IRoutineAnalysis.providedBy(obj):
-        request = obj.getRequest()
-        if request:
-            request.reindexObject()
+def _reindex_request(obj, idxs=None):
+    if not IRequestAnalysis.providedBy(obj):
```
Comment (ramonski) [inline bika/lims/workflow/analysis/events.py]: What's the difference between `IRequestAnalysis` and `IAnalysis`?

Comment (xispa) [reply bika/lims/workflow/analysis/events.py]: `IRequestAnalysis` interface is implemented by those analyses that belong, directly or indirectly, to an `AnalysisRequest`. At the moment, `Analysis` (directly), `DuplicateAnalysis` (indirectly). As such, thsese objects implement the function `getRequest` (and related).

`ReferenceAnalysis` objects are not associated to `AnalysisRequest`, rather their parent is directly a `Worksheet`, so they don't implement the function `getRequest` and related.

**Comment by xispa** [comment]
Re `getCurrentStatus` vs `api.get_workflow_status_of`. `bika.lims.workflow` contains some functions that are useful for workflow stuff, and others will come. I suggest to create package for `api`, move the functions from `api.py` to `__init__.py`, create a `workflow.py` file inside that package and add the workflow related functions there. Then, we'll be able to do something like `bika.lims.api.workflow.get_workflow_status_of`:

```
lims
 |_ api
     |_ __init__.py
     |_ workflow.py
     |_ ...
```

**Comment by ramonski** [comment]
Yes, that was actually the plan with `senaite.api`, see here: https://github.com/senaite/senaite.api/blob/master/src/senaite/api/__init__.py
So I totally agree with that approach.

However, I'm not sure when the right moment is there to switch completely to `senaite.api` as a dependency of `senaite.core`. It would allow us to have separate releases and maintain this package outside of the core. For us developers that would just mean a separate entry in `mr.developer` and to release a new version if something changed.

**Comment by xispa** [comment]
Agree. Will then follow this approach. In my opinion, we'll be able to switch to `senaite.api`as soon as we have this workflow thing there. What is important here is to ensure `senaite.api` remains totally decoupled of `senaite.core`, without depending on it.

File: bika/lims/workflow/analysis/events.py (hunk 4 above)
Comment (ramonski) [reply bika/lims/workflow/analysis/events.py]: Thanks for the explanation Jordi

File: bika/lims/workflow/__init__.py [hunk 5]
```diff
@@ -370,13 +370,12 @@ def getReviewHistory(instance):
...
 
-
+@deprecated('Use api.get_workflow_status_of instead')
 def getCurrentState(obj, stateflowid='review_state'):
     """ The current state of the object for the state flow id specified
         Return empty if there's no workflow state for the object and flow id
     """
-    wf = getToolByName(obj, 'portal_workflow')
-    return wf.getInfoFor(obj, stateflowid, '')
+    return api.get_workflow_status_of(obj, stateflowid)
```
Comment (ramonski) [inline bika/lims/workflow/__init__.py]: cool

**Pull request merged**